internal/freelist/freelist.go

```diff
@@ -0,0 +1,86 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type Freelist interface {
+	// Init initializes this freelist with the given list of pages.
+	Init(ids common.Pgids)
+
+	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
+	// If a contiguous block cannot be found then 0 is returned.
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// Release moves all page ids for a transaction id (or older) to the freelist.
+	Release(txId common.Txid)
+
+	// ReleaseRange moves pending pages allocated within an extent [begin,end] to the free list.
+	ReleaseRange(begin, end common.Txid)
+
+	// Free releases a page and its overflow for a given transaction id.
+	// If the page is already free then a panic will occur.
+	Free(txId common.Txid, p *common.Page)
+
+	// Freed returns whether a given page is in the free list.
+	Freed(pgId common.Pgid) bool
+
+	// Rollback removes the pages from a given pending tx.
+	Rollback(txId common.Txid)
+
+	// pendingPageIds returns all pending pages by transaction id.
+	pendingPageIds() map[common.Txid]*txPending
```

Comment:
It's just an internal method, no need to add it into the interface.
```suggestion
```

Reply:
you need it in the interface for Copyall & NoSyncReload